front/lib/api/assistant/actions/conversation/include_file.ts

```diff
@@ -0,0 +1,171 @@
+import logger from "@app/logger/logger";
+
+export function isConversationIncludableFileContentType(
+  contentType: SupportedContentFragmentType
+): boolean {
+  if (isSupportedImageContentType(contentType)) {
+    return false;
+  }
+  // For now we only allow including text files.
+  switch (contentType) {
+    case "application/msword":
+    case "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
+    case "application/pdf":
+    case "text/markdown":
+    case "text/plain":
+    case "dust-application/slack":
+      return true;
+
+    case "text/comma-separated-values":
+    case "text/csv":
+    case "text/tab-separated-values":
+    case "text/tsv":
+      return false;
+    default:
+      assertNever(contentType);
+  }
+}
+
+interface ConversationIncludeFilesActionBlob {
+  id: ModelId;
+  agentMessageId: ModelId;
+  params: {
+    fileId: string;
+  };
+  functionCallId: string | null;
+  functionCallName: string | null;
+  step: number;
+}
+
+const CONTEXT_SIZE_DIVISOR_FOR_INCLUDE = 4;
+
+export class ConversationIncludeFilesAction extends BaseAction {
+  readonly agentMessageId: ModelId;
+  readonly params: {
+    fileId: string;
+  };
+  readonly contentFragments: null[] = [];
+  readonly functionCallId: string | null;
+  readonly functionCallName: string | null;
+  readonly step: number = -1;
+  readonly type = "conversation_include_files_action";
+
+  constructor(blob: ConversationIncludeFilesActionBlob) {
+    super(blob.id, "conversation_include_files_action");
+
+    this.agentMessageId = blob.agentMessageId;
+    this.params = blob.params;
+    this.functionCallId = blob.functionCallId;
+    this.functionCallName = blob.functionCallName;
+    this.step = blob.step;
+  }
+
+  renderForFunctionCall(): FunctionCallType {
+    return {
+      id: this.functionCallId ?? `call_${this.id.toString()}`,
+      name: this.functionCallName ?? "include_conversation_files",
+      arguments: JSON.stringify(this.params),
+    };
+  }
+
+  async renderForMultiActionsModel({
+    conversation,
+    model,
+  }: {
+    conversation: ConversationType;
+    model: ModelConfigurationType;
+  }): Promise<FunctionMessageTypeModel> {
+    const finalize = (content: string) => {
+      return {
+        role: "function" as const,
+        name: this.functionCallName ?? "include_conversation_files",
+        function_call_id: this.functionCallId ?? `call_${this.id.toString()}`,
+        content,
+      };
+    };
+
+    const m = (conversation.content.flat(1).find((m) => {
+      if (
+        isContentFragmentType(m) &&
+        isConversationIncludableFileContentType(m.contentType) &&
+        isConversationIncludableFileContentType(m.contentType) &&
+        m.fileId === this.params.fileId
+      ) {
+        return true;
+      }
+      return false;
+    }) || null) as ContentFragmentType | null;
+
+    if (!m) {
+      return finalize(
+        `Error: File \`${this.params.fileId}\` not found in conversation`
```

Comment:
note:
We also expose in the list, the files produced by the output of tools (table-query for now but we can imagine other soon, such as extract).
I believe going thru the fragments only will not see it.

Maybe it's worth sharing getting the list of files from the same code than the list files action and then filter with `isConversationIncludableFileContentType()`. (i understand we'll lose the early return but on the flip side, we avoid a sneaky "bug" in the future).

Reply:
Actually will keep that for subsequent PRs to limit the refactor here since it is already getting big. But completely aligned.